# Heaps

## 1. Heap Fundamentals

A heap is a complete binary tree stored in an array. The **heap-order property** requires that every parent node compares favorably with its children: smaller in a min-heap, larger in a max-heap. This guarantees the extreme value (min or max) is always at the root.

- **Min-heap:** every node's value <= its children's values. Smallest element at root.
- **Max-heap:** every node's value >= its children's values. Largest element at root.
- **Complete binary tree:** all levels full except possibly the last, which fills left to right.
- **Array storage:** parent-child relationships computed from indices, no pointers needed.

### Terminology

| Term | Definition |
|------|-----------|
| Heap-order property | Every parent <= its children (min-heap) or >= its children (max-heap), so the root is always the extreme value |
| Complete binary tree | All levels full except possibly the last, filled left to right |
| Sift up | After `H.push(val)`, move the new element upward by swapping with its parent until heap order holds |
| Sift down | After `H.pop()`, move the replacement root downward by swapping with its smallest (or largest) child until heap order holds |
| Heapify | `heapify(arr)` -- convert an unordered array into a heap in O(n) by sifting down from the last non-leaf to the root |

### Array Storage

For a node at index `i` (0-indexed):

| Relationship | Formula |
|---|---|
| Left child | `2i + 1` |
| Right child | `2i + 2` |
| Parent | `(i - 1) // 2` |
| Last non-leaf | `(n - 2) // 2` |

```text
        1
       / \
      3    5
     / \
    7   9

array: [1, 3, 5, 7, 9]
index:  0  1  2  3  4

parent(3) = (3-1)//2 = 1 -> val 3
left(1)   = 2*1+1    = 3 -> val 7
right(1)  = 2*1+2    = 4 -> val 9
```

### Properties

| Property | Detail |
|---|---|
| **Key idea** | Complete binary tree with heap-order property, stored as array |
| **Use-case** | Priority queues, top-k, scheduling, graph algorithms (Dijkstra, Prim) |
| **Time** | O(log n) insert/extract, O(1) peek, O(n) build |
| **Space** | O(n) |

### Operations

| Operation | Description | Time |
|-----------|-------------|------|
| `H.push(val)` | Insert val, sift up to restore order | O(log n) |
| `H.pop()` | Remove and return root, sift down to restore order | O(log n) |
| `H.peek()` | Return root without removing | O(1) |
| `heapify(arr)` | Build heap from unordered array (bottom-up sift down) | O(n) |

## 2. Min Heap

A min-heap keeps the smallest element at the root.

- **Insert (push):** append to end of array, sift up by swapping with parent while smaller
- **Extract min (pop):** swap root with last element, remove last, sift down root by swapping with smallest child while larger
- **Peek:** return `heap[0]`

In [1]:
class MinHeap:
    def __init__(self):
        self.heap = []

    def push(self, val):
        self.heap.append(val)
        self._siftUp(len(self.heap) - 1)

    def pop(self):
        if not self.heap:
            return None
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        val = self.heap.pop()
        if self.heap:
            self._siftDown(0)
        return val

    def peek(self):
        return self.heap[0] if self.heap else None

    def _siftUp(self, i):
        while i > 0:
            parent = (i - 1) // 2
            if self.heap[i] < self.heap[parent]:
                self.heap[i], self.heap[parent] = self.heap[parent], self.heap[i]
                i = parent
            else:
                break

    def _siftDown(self, i):
        n = len(self.heap)
        while 2 * i + 1 < n:
            left = 2 * i + 1
            right = 2 * i + 2
            smallest = i
            if left < n and self.heap[left] < self.heap[smallest]:
                smallest = left
            if right < n and self.heap[right] < self.heap[smallest]:
                smallest = right
            if smallest != i:
                self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
                i = smallest
            else:
                break

    def __len__(self):
        return len(self.heap)

#### Steps

`H = MinHeap()`, push 5, 3, 8, 1, 4.

| Step | Operation | Heap (array) | Notes |
|------|-----------|-------------|-------|
| 1 | `H.push(5)` | `[5]` | single element |
| 2 | `H.push(3)` | `[3, 5]` | 3 < 5, sift up to root |
| 3 | `H.push(8)` | `[3, 5, 8]` | 8 > 3, stays |
| 4 | `H.push(1)` | `[1, 3, 8, 5]` | 1 sifts up past 5 and 3 to root |
| 5 | `H.push(4)` | `[1, 3, 8, 5, 4]` | 4 > 3, stays |
| 6 | `H.peek()` | `[1, 3, 8, 5, 4]` | returns 1 |
| 7 | `H.pop()` | `[3, 4, 8, 5]` | returns 1, sift 4 down past 3 |
| 8 | `H.pop()` | `[4, 5, 8]` | returns 3, sift 5 down past 4 |

```text
after step 5:

        1
       / \
      3    8
     / \
    5   4
```

## 3. Max Heap

A max-heap keeps the largest element at the root. Structurally identical to a min-heap with reversed comparisons in sift operations.

In [2]:
class MaxHeap:
    def __init__(self):
        self.heap = []

    def push(self, val):
        self.heap.append(val)
        self._siftUp(len(self.heap) - 1)

    def pop(self):
        if not self.heap:
            return None
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        val = self.heap.pop()
        if self.heap:
            self._siftDown(0)
        return val

    def peek(self):
        return self.heap[0] if self.heap else None

    def _siftUp(self, i):
        while i > 0:
            parent = (i - 1) // 2
            if self.heap[i] > self.heap[parent]:
                self.heap[i], self.heap[parent] = self.heap[parent], self.heap[i]
                i = parent
            else:
                break

    def _siftDown(self, i):
        n = len(self.heap)
        while 2 * i + 1 < n:
            left = 2 * i + 1
            right = 2 * i + 2
            largest = i
            if left < n and self.heap[left] > self.heap[largest]:
                largest = left
            if right < n and self.heap[right] > self.heap[largest]:
                largest = right
            if largest != i:
                self.heap[i], self.heap[largest] = self.heap[largest], self.heap[i]
                i = largest
            else:
                break

    def __len__(self):
        return len(self.heap)

#### Steps

`H = MaxHeap()`, push 3, 5, 1, 8.

| Step | Operation | Heap (array) | Notes |
|------|-----------|-------------|-------|
| 1 | `H.push(3)` | `[3]` | single element |
| 2 | `H.push(5)` | `[5, 3]` | 5 > 3, sift up to root |
| 3 | `H.push(1)` | `[5, 3, 1]` | 1 < 5, stays |
| 4 | `H.push(8)` | `[8, 5, 1, 3]` | 8 sifts up past 3 and 5 to root |
| 5 | `H.pop()` | `[5, 3, 1]` | returns 8, sift 3 down: 3 < 5, swap |
| 6 | `H.pop()` | `[3, 1]` | returns 5, sift 1 down: 1 < 3, swap |

## 4. Heapify (Build Heap)

Bottom-up heapify converts an unordered array into a heap in O(n).

- Start from the last non-leaf node: index `(n - 2) // 2`
- Sift each node down to its correct position
- Nodes near the bottom have short sift paths, and most nodes are near the bottom
- Total work is O(n): level k has up to $2^k$ nodes, each sifts at most $h - k$ levels

In [3]:
def _siftDown(arr, i, n):
    while 2 * i + 1 < n:
        left = 2 * i + 1
        right = 2 * i + 2
        smallest = i
        if left < n and arr[left] < arr[smallest]:
            smallest = left
        if right < n and arr[right] < arr[smallest]:
            smallest = right
        if smallest != i:
            arr[i], arr[smallest] = arr[smallest], arr[i]
            i = smallest
        else:
            break


def heapify(arr):
    n = len(arr)
    for i in range((n - 2) // 2, -1, -1):
        _siftDown(arr, i, n)


nums = [9, 5, 6, 2, 3]
heapify(nums)
print(nums)

[2, 3, 6, 5, 9]


#### Steps

`heapify([9, 5, 6, 2, 3])` (min-heap).

```text
initial:

        9
       / \
      5    6
     / \
    2   3

last non-leaf: (5-2)//2 = 1 (val 5)
```

| Step | siftDown(i) | Array before | Swap | Array after |
|------|-------------|-------------|------|-------------|
| 1 | i=1 (val 5) | `[9, 5, 6, 2, 3]` | 5 > 2, swap(1,3) | `[9, 2, 6, 5, 3]` |
| 2 | i=0 (val 9) | `[9, 2, 6, 5, 3]` | 9 > 2, swap(0,1) | `[2, 9, 6, 5, 3]` |
| 3 | continue i=1 | `[2, 9, 6, 5, 3]` | 9 > 3, swap(1,4) | `[2, 3, 6, 5, 9]` |

```text
result:

        2
       / \
      3    6
     / \
    5   9
```

## 5. Heap Sort

Heap sort builds a max-heap, then repeatedly swaps the root (maximum) with the last unsorted element and sifts down.

- Phase 1: build a max-heap from the array in O(n)
- Phase 2: extract maximum n-1 times, placing each at the end in O(n log n)
- Sorts ascending because the largest elements accumulate at the back

| Property | Detail |
|---|---|
| **Key idea** | Build max-heap, repeatedly extract max to end of array |
| **Time** | O(n log n) all cases |
| **Space** | O(1) in-place |
| **Stable** | No |

In [4]:
def _siftDownMax(arr, i, n):
    while 2 * i + 1 < n:
        left = 2 * i + 1
        right = 2 * i + 2
        largest = i
        if left < n and arr[left] > arr[largest]:
            largest = left
        if right < n and arr[right] > arr[largest]:
            largest = right
        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            i = largest
        else:
            break


def heapSort(nums):
    n = len(nums)
    for i in range((n - 2) // 2, -1, -1):
        _siftDownMax(nums, i, n)
    for i in range(n - 1, 0, -1):
        nums[0], nums[i] = nums[i], nums[0]
        _siftDownMax(nums, 0, i)
    return nums


print(heapSort([4, 1, 7, 3, 2]))

[1, 2, 3, 4, 7]


#### Steps

`heapSort([4, 1, 7, 3, 2])`

**Phase 1: build max-heap**

| Step | siftDown(i) | Swap | Array |
|------|-------------|------|-------|
| 1 | i=1 (val 1) | 1 < 3, swap(1,3) | `[4, 3, 7, 1, 2]` |
| 2 | i=0 (val 4) | 4 < 7, swap(0,2) | `[7, 3, 4, 1, 2]` |

Max-heap: `[7, 3, 4, 1, 2]`

**Phase 2: extract and sort**

Elements after the swap position are sorted and fixed.

| Step | Swap | siftDown | Array |
|------|------|----------|-------|
| 1 | swap(0,4) | 2 < 4, swap(0,2) | `[4, 3, 2, 1, 7]` |
| 2 | swap(0,3) | 1 < 3, swap(0,1) | `[3, 1, 2, 4, 7]` |
| 3 | swap(0,2) | no swap needed | `[2, 1, 3, 4, 7]` |
| 4 | swap(0,1) | -- | `[1, 2, 3, 4, 7]` |

Result: `[1, 2, 3, 4, 7]`

## 6. Python heapq Module

Python's `heapq` module provides a min-heap backed by a regular list. No separate heap object -- operates directly on a plain list.

| Operation | Description | Time |
|-----------|-------------|------|
| `heapq.heappush(h, val)` | Push val onto heap | O(log n) |
| `heapq.heappop(h)` | Pop and return smallest | O(log n) |
| `heapq.heappushpop(h, val)` | Push val, then pop smallest | O(log n) |
| `heapq.heapreplace(h, val)` | Pop smallest, then push val | O(log n) |
| `heapq.heapify(x)` | Transform list into heap in-place | O(n) |
| `heapq.nlargest(k, iterable)` | Return k largest elements | O(n log k) |
| `heapq.nsmallest(k, iterable)` | Return k smallest elements | O(n log k) |
| `h[0]` | Peek at smallest without removing | O(1) |

In [5]:
import heapq

h = []
heapq.heappush(h, 5)
heapq.heappush(h, 3)
heapq.heappush(h, 8)
heapq.heappush(h, 1)
print(h[0])
print(heapq.heappop(h))
print(heapq.heappop(h))

nums = [9, 5, 6, 2, 3]
heapq.heapify(nums)
print(nums)

print(heapq.nlargest(3, [4, 1, 7, 3, 2]))
print(heapq.nsmallest(2, [4, 1, 7, 3, 2]))

1
1
3
[2, 3, 6, 5, 9]
[7, 4, 3]
[1, 2]


### Max-Heap with heapq (Negation Trick)

`heapq` only provides a min-heap. To simulate a max-heap, negate values on push and negate again on pop.

In [10]:
maxH = []
for val in [3, 1, 5, 2]:
    heapq.heappush(maxH, -val)

print(-heapq.heappop(maxH))
print(-heapq.heappop(maxH))

tasks = []
heapq.heappush(tasks, (-5, "urgent"))
heapq.heappush(tasks, (-1, "low"))
priority, name = heapq.heappop(tasks)
print(-priority, name)

5
3
5 urgent


### Custom Comparison

`heapq` compares elements with `<`. For tuples, comparison is lexicographic.

- To heap custom objects, wrap in `(priority, tiebreaker, obj)` tuples
- An incrementing counter as tiebreaker prevents comparing uncomparable objects and preserves insertion order

In [11]:
counter = 0
pq = []
for task, pri in [("email", 2), ("build", 1), ("test", 1)]:
    heapq.heappush(pq, (pri, counter, task))
    counter += 1

while pq:
    pri, _, task = heapq.heappop(pq)
    print(pri, task)

1 build
1 test
2 email


## 7. Patterns

| Pattern | Description | Problem keywords |
|---------|-------------|-----------------|
| **Top-k elements** | Maintain a size-k min-heap, pop when size exceeds k | "k largest", "k most frequent", "k closest" |
| **Kth largest/smallest** | Root of a size-k min/max-heap is the answer | "kth largest", "kth smallest" |
| **Merge k sorted** | Min-heap of size k, pop smallest, push next from same source | "merge k sorted lists/arrays" |
| **Two heaps (median)** | Max-heap for lower half, min-heap for upper half | "median from data stream", "sliding median" |
| **Greedy with heap** | Pop min/max to make greedy choices | "minimum cost", "task scheduler", "reorganize string" |

All pattern implementations below reuse the `MinHeap` and `MaxHeap` classes from sections 2-3.

### Top-K Elements

Maintain a min-heap of size k. For each element: push onto heap, pop smallest if size exceeds k. After processing, the heap contains the k largest elements. The root is the kth largest.

In [6]:
def topKLargest(nums, k):
    heap = MinHeap()
    for val in nums:
        heap.push(val)
        if len(heap) > k:
            heap.pop()
    res = []
    while len(heap):
        res.append(heap.pop())
    return res[::-1]


print(topKLargest([3, 1, 5, 12, 2, 11], 3))

[12, 11, 5]


#### Steps

`topKLargest([3, 1, 5, 12, 2, 11], k=3)`

| Step | val | Heap (after push) | Size > k? | Heap (after pop) |
|------|-----|-------------------|-----------|-----------------|
| 1 | 3 | `[3]` | no | `[3]` |
| 2 | 1 | `[1, 3]` | no | `[1, 3]` |
| 3 | 5 | `[1, 3, 5]` | no | `[1, 3, 5]` |
| 4 | 12 | `[1, 3, 5, 12]` | yes, pop 1 | `[3, 12, 5]` |
| 5 | 2 | `[2, 3, 5, 12]` | yes, pop 2 | `[3, 12, 5]` |
| 6 | 11 | `[3, 11, 5, 12]` | yes, pop 3 | `[5, 11, 12]` |

Result: `[12, 11, 5]`

### Kth Largest Element

Same as top-k, but return only the root of the size-k heap. O(n log k).

Alternative: heapify the full array as a max-heap, then pop k times. O(n + k log n). Better when k is small relative to n.

In [7]:
def kthLargest(nums, k):
    heap = MinHeap()
    for val in nums:
        heap.push(val)
        if len(heap) > k:
            heap.pop()
    return heap.peek()


print(kthLargest([3, 2, 1, 5, 6, 4], 2))

5


### Merge K Sorted Lists

Push the first element from each list as `(val, listIdx, elemIdx)`. Pop the smallest, append to result, push the next element from the same list. Heap size never exceeds k (number of lists).

In [8]:
def mergeKSorted(lists):
    heap = MinHeap()
    for i, lst in enumerate(lists):
        if lst:
            heap.push((lst[0], i, 0))
    res = []
    while len(heap):
        val, li, ei = heap.pop()
        res.append(val)
        if ei + 1 < len(lists[li]):
            heap.push((lists[li][ei + 1], li, ei + 1))
    return res


print(mergeKSorted([[1, 4, 5], [1, 3, 4], [2, 6]]))

[1, 1, 2, 3, 4, 4, 5, 6]


#### Steps

`mergeKSorted([[1, 4, 5], [1, 3, 4], [2, 6]])`

n = total elements across all lists, k = number of lists.

| Step | Pop | Push next | Heap | Result |
|------|-----|-----------|------|--------|
| init | -- | (1,0,0), (1,1,0), (2,2,0) | `[(1,0,0), (1,1,0), (2,2,0)]` | `[]` |
| 1 | (1,0,0) | (4,0,1) | `[(1,1,0), (2,2,0), (4,0,1)]` | `[1]` |
| 2 | (1,1,0) | (3,1,1) | `[(2,2,0), (4,0,1), (3,1,1)]` | `[1, 1]` |
| 3 | (2,2,0) | (6,2,1) | `[(3,1,1), (4,0,1), (6,2,1)]` | `[1, 1, 2]` |
| 4 | (3,1,1) | (4,1,2) | `[(4,0,1), (6,2,1), (4,1,2)]` | `[1, 1, 2, 3]` |
| 5 | (4,0,1) | (5,0,2) | `[(4,1,2), (6,2,1), (5,0,2)]` | `[1, 1, 2, 3, 4]` |
| 6 | (4,1,2) | -- | `[(5,0,2), (6,2,1)]` | `[1, 1, 2, 3, 4, 4]` |
| 7 | (5,0,2) | -- | `[(6,2,1)]` | `[1, 1, 2, 3, 4, 4, 5]` |
| 8 | (6,2,1) | -- | `[]` | `[1, 1, 2, 3, 4, 4, 5, 6]` |

Each element pushed and popped once: O(n log k) total.

### Running Median (Two Heaps)

Maintain two heaps:
- `lo`: max-heap (negated) for the lower half of values
- `hi`: min-heap for the upper half of values

Invariant: `len(lo) == len(hi)` or `len(lo) == len(hi) + 1`.
- Odd count: median is root of `lo`
- Even count: median is average of both roots

In [9]:
class MedianFinder:
    def __init__(self):
        self.lo = MaxHeap()
        self.hi = MinHeap()

    def addNum(self, num):
        self.lo.push(num)
        self.hi.push(self.lo.pop())
        if len(self.hi) > len(self.lo):
            self.lo.push(self.hi.pop())

    def findMedian(self):
        if len(self.lo) > len(self.hi):
            return self.lo.peek()
        return (self.lo.peek() + self.hi.peek()) / 2


mf = MedianFinder()
for val in [5, 2, 8, 1, 4]:
    mf.addNum(val)
    print(f"add {val} -> median = {mf.findMedian()}")

add 5 -> median = 5
add 2 -> median = 3.5
add 8 -> median = 5
add 1 -> median = 3.5
add 4 -> median = 4


#### Steps

`MedianFinder()`, add 5, 2, 8, 1, 4.

| Step | addNum | lo (max-heap) | hi (min-heap) | Median |
|------|--------|--------------|---------------|--------|
| 1 | 5 | `[5]` | `[]` | 5 |
| 2 | 2 | `[2]` | `[5]` | (2 + 5) / 2 = 3.5 |
| 3 | 8 | `[5, 2]` | `[8]` | 5 |
| 4 | 1 | `[2, 1]` | `[5, 8]` | (2 + 5) / 2 = 3.5 |
| 5 | 4 | `[4, 1, 2]` | `[5, 8]` | 4 |

`lo` is a max-heap holding the smaller half. `hi` is a min-heap holding the larger half.
The root of `lo` is always the largest value in the lower half.

## 8. Reference

### Operation Complexity

| Operation | Time | Notes |
|-----------|------|-------|
| Push (insert) | O(log n) | Sift up at most h levels where h = floor(log n) |
| Pop (extract) | O(log n) | Sift down at most h levels |
| Peek | O(1) | Root is always min or max |
| Heapify (build) | O(n) | Bottom-up: deeper levels have more nodes but shorter sift paths |
| Search | O(n) | No ordering beyond parent-child relationship |
| Delete arbitrary | O(n) | O(n) search + O(log n) sift |
| Decrease key | O(n) or O(log n) | O(n) to find the element, O(log n) to sift up. O(log n) total if index is tracked via a hash map. Critical for Dijkstra and Prim. |

### Heap vs Other Structures

| Structure | Insert | Access min/max | Delete min/max | Search | Sorted |
|-----------|--------|---------------|---------------|--------|--------|
| Heap | O(log n) | O(1) | O(log n) | O(n) | No |
| Sorted array | O(n) | O(1) | O(1) or O(n) | O(log n) | Yes |
| BST (balanced) | O(log n) | O(log n) | O(log n) | O(log n) | Yes |
| Unsorted array | O(1) | O(n) | O(n) | O(n) | No |

### Pattern Quick Reference

| Pattern | Heap type | Heap size | Time | Use-case | Problem keywords |
|---------|-----------|-----------|------|----------|-----------------|
| Top-k largest | min-heap of size k | k | O(n log k) | Find the k biggest elements from a stream or array | "k largest", "k most frequent", "k closest points" |
| Top-k smallest | max-heap of size k | k | O(n log k) | Find the k smallest elements | "k smallest", "k weakest rows" |
| Kth largest | min-heap of size k | k | O(n log k) | Return a single rank-based element | "kth largest", "kth smallest", "kth closest" |
| Kth largest (alt) | max-heap of full array | n | O(n + k log n) | Better when k is small relative to n | "kth largest element" |
| Merge k sorted | min-heap | k | O(n log k) | Combine multiple ordered sources into one | "merge k sorted lists", "smallest range covering k lists" |
| Running median | two heaps | n | O(n log n) | Track median as values arrive one at a time | "find median from data stream", "sliding window median" |
| Heap sort | max-heap in-place | n | O(n log n) | In-place O(n log n) sort with no extra arrays | "sort an array" (guaranteed worst-case) |

### Cross-References

- Heap sort comparison with other sorts: see `sorting.ipynb`
- Priority queue basics: see `stacks_queues.ipynb`
- Dijkstra's shortest path: uses min-heap as priority queue with decrease-key